usr.sbin/inetd/parse_v2.c

```diff
@@ -0,0 +1,1067 @@
+		is_valid = setup_internal(sep) && is_valid;
+	}
+
+	if (sep->se_socktype == SE_SOCKTYPE_UNINIT) {
+		/* Ensure socktype is specified (either set or inferred) */
+		ENI("socktype");
+		is_valid = false;
+	}
+
+	if (sep->se_wait == SE_WAIT_UNINIT) {
+		/* Ensure wait is specified */
+		ENI("wait");
+		is_valid = false;
+	}
+
+	if (sep->se_user == NULL) {
+		/* Ensure user is specified */
+		ENI("user");
+		is_valid = false;
+	}
+
+	if (sep->se_proto == NULL) {
+		/* Ensure protocol is specified */
+		ENI("protocol");
+		is_valid = false;
+	} else {
+		is_valid = infer_protocol_ip_version(sep) && is_valid;
+	}
+
+#ifdef IPSEC
+	setup_ipsec(sep);
+#endif
+	return is_valid;
+}
+
+/* fill_default_values related functions */
+#ifdef IPSEC
+static void
+setup_ipsec(struct servtab * sep)
+{
+	if (sep->se_policy == NULL) {
+		/* Set to default global policy */
+		sep->se_policy = policy;
+	} else if (strlen(sep->se_policy) == 0) {
+		/* IPsec was intentionally disabled. */
+		free(sep->se_policy);
+		sep->se_policy = NULL;
+	}
+}
+#endif
+
+static void
+try_infer_socktype(struct servtab * sep) {
+	if (sep->se_socktype == SE_SOCKTYPE_UNINIT) {
+		/* Check values of se_proto udp, udp6, tcp, tcp6 to set dgram/stream */
+		if (sep->se_proto != NULL) {
+			if (strncmp(sep->se_proto, "udp", 3) == 0) {
+				sep->se_socktype = SOCK_DGRAM;
+			} else if (strncmp(sep->se_proto, "tcp", 3) == 0) {
+				sep->se_socktype = SOCK_STREAM;
+			}
+		}
+	}
+}
+
+static bool
+setup_internal(struct servtab * sep)
+{
+	pid_t wait_prev = sep->se_wait;
+	if (parse_server(sep, "internal") != 0) {
+		ENI("exec");
+		return false;
+	} else if (wait_prev != SE_WAIT_UNINIT && wait_prev != sep->se_wait) {
+		/* If wait was already specified throw an error. */
+		WRN(WAIT_WRN, sep->se_service, (sep->se_wait ? "yes" : "no"));
+	}
+	return true;
+}
+
+static bool
+infer_protocol_ip_version(struct servtab *sep)
+{
+	struct in_addr tmp;
+
+	if (strcmp("tcp", sep->se_proto) == 0
+		|| strcmp("udp", sep->se_proto) == 0
+		|| strcmp("rpc/tcp", sep->se_proto) == 0
+		|| strcmp("rpc/udp", sep->se_proto) == 0) {
+
+		if (inet_pton(AF_INET6, sep->se_hostaddr, &tmp)) {
+			sep->se_family = AF_INET6;
+		} else if (!inet_pton(AF_INET, sep->se_hostaddr, &tmp)) {
+			ERR("Address family of %s is ambigous or invalid. "
+				"Explicitly specify protocol", sep->se_hostaddr);
+			return false;
+		}
+	}
+	
+	return true;
+}
+
+/* 
+ * Skips whitespaces, newline characters, and comments, 
+ * and returns the next token. Returns false and logs error if an EOF is 
+ * encountered. 
+ */
+static bool 
+skip_whitespace(char **cpp) {
+	char *cp = *cpp;
+
+	int line_start = line_number;
+	
+	for (;;) {
+		while (isblank((int)*cp))
+			cp++;
+
+		if (*cp == '\0' || *cp == '#') {  
+			cp = nextline(fconfig);
+
+			/* Should never expect EOF when skipping whitespace */
+			if (cp == NULL) {
+				ERR("Early end of file after line %d", line_start); 
+				return false;
+			}
+
+		} else break;
```

Comment:
early returns again; if you "continue;" in the if part then you don't need an else.